src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -875,27 +887,62 @@ private bool IsAttributeDeclarationTerminator()
+                this.CurrentNodeKind == SyntaxKind.AttributeList &&
+                !inExpressionContext)
             {
                 return (AttributeListSyntax)this.EatNode();
             }
 
+            // May have to reset if we discover this is not an attribute but is instead a collection literal.
+            using var resetPoint = GetDisposableResetPoint(resetOnDispose: false);
+
             var openBracket = this.EatToken(SyntaxKind.OpenBracketToken);
 
             // Check for optional location :
             var location = IsSomeWord(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken
                 ? _syntaxFactory.AttributeTargetSpecifier(ConvertToKeyword(this.EatToken()), this.EatToken(SyntaxKind.ColonToken))
                 : null;
 
-            var attributes = _pool.AllocateSeparated<AttributeSyntax>();
-            this.ParseAttributes(attributes);
-            return _syntaxFactory.AttributeList(
-                openBracket,
-                location,
-                _pool.ToListAndFree(attributes),
-                this.EatToken(SyntaxKind.CloseBracketToken));
+            var attributesBuilder = _pool.AllocateSeparated<AttributeSyntax>();
+            this.ParseAttributes(attributesBuilder);
+            var attributes = _pool.ToListAndFree(attributesBuilder);
+
+            var closeBracket = this.EatToken(SyntaxKind.CloseBracketToken);
+            if (inExpressionContext && ParseAsCollectionLiteral())
+            {
+                // we're in an expression and we've seen `[A, B].`  This is actually the start of a collection literal
+                // that someone is explicitly accessing a member off of.
+                resetPoint.Reset();
+                return null;
+            }
+
+            return _syntaxFactory.AttributeList(openBracket, location, attributes, closeBracket);
+
+            bool ParseAsCollectionLiteral()
+            {
+                // `[A, B].` is a member access off of a collection literal. 
+                if (this.CurrentToken.Kind == SyntaxKind.DotToken)
+                    return true;
+
+                // `[A, B]->` is a member access off of a collection literal. Note: this will always be illegal
+                // semantically (as a collection literal has the natural type List<> which is not a pointer type).  But
+                // we leave that check to binding.
+                if (this.CurrentToken.Kind == SyntaxKind.MinusGreaterThanToken)
+                    return true;
+
+                // `[A, B]?.`  The `?` is unnecessary (as a collection literal is always non-null).  But it is fine if
+                // the use wants to include it unnecessarily.  A linter could always advise they remove it.
+                if (this.CurrentToken.Kind == SyntaxKind.QuestionToken &&
+                    this.PeekToken(1).Kind == SyntaxKind.DotToken)
+                {
+                    return true;
+                }
```

Comment:
>} [](http://example.com/codeflow?start=16&length=1)

Do we need to handle `?[` as well? `[A, B]?[C].ToString();`

Reply:
so this is tracked with the general test bullet point about ambiguities with ternary and conditional indexing.

Reply:
Another case: should this be parsed as a cast or indexer access? `(name)[a,b,c]`

Reply:
Great case. @jnm2